module/mfrc522_parser.c

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include "linux/kernel.h"
+#include "mfrc522_command.h"
+
+#include <linux/string.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+
+#include "mfrc522_parser.h"
+
+#define MFRC522_SEPARATOR ":"
+#define MFRC522_CMD_AMOUNT 4
+
+struct command {
+	const char *input;
+	u8 parameter_amount;
+	u8 cmd;
+};
+
+static struct command commands[MFRC522_CMD_AMOUNT] = {
+	{ .input = "mem_write",
+	  .parameter_amount = 2,
+	  .cmd = MFRC522_CMD_MEM_WRITE },
+	{ .input = "mem_read",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_MEM_READ },
+	{ .input = "gen_rand_id",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_GEN_RANDOM },
+	{ .input = "version",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_GET_VERSION },
+};
+
+/**
+ * Get the command associated with a command name
+ *
+ * @param token Command name found during parsing
+ *
+ * @return Return a pointer to a command declared in the commands array, or NULL if no
+ *         command matched the input
+ */
+static struct command *cmd_from_token(const char *token)
+{
+	size_t i;
+
+	for (i = 0; i < MFRC522_CMD_AMOUNT; i++)
+		if (!strcmp(token, commands[i].input))
+			return &commands[i];
+
+	return NULL;
+}
+
+/**
+ * Parse a command with multiple arguments
+ *
+ * @param input Remaining user input, mutable and allocated via kmalloc() in mfrc522_parse()
+ * @param cmd Command returned by cmd_from_token()
+ *
+ * @return A complete MFRC522 command with allocated memory for the extra data
+ */
+static struct mfrc522_command *parse_multi_arg(char *input, struct command *cmd)
+{
+	// We only enter this function if arguments have been given to the input
+	u8 parameter_amount = 0;
+	char *token;
+	char *extra_data;
+	u8 extra_data_len;
+	int ret;
+
+	while (input) {
+		token = strsep(&input, MFRC522_SEPARATOR);
+		parameter_amount++;
+
+		// The first parameter is the extra data
+		if (parameter_amount == 1)
+			extra_data = kstrdup(token, GFP_KERNEL);
+
+		// The second parameter is the length of the extr data
+		if (parameter_amount == 2) {
+			ret = kstrtou8(token, 10, &extra_data_len);
+			if (ret == -EINVAL) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected number but got %s\n",
+				       token);
+				goto err;
+			}
+
+			if (ret == -ERANGE) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected Unsigned Byte (0 - 255) but got %s\n",
+				       token);
+				goto err;
+			}
+		}
+	}
+
+	pr_info("[MFRC522] Found %d parameters for command %s\n",
+		parameter_amount, cmd->input);
+
+	if (parameter_amount != cmd->parameter_amount) {
+		pr_err("[MFRC522] Invalid command: %s: Expected %d arguments but got %d\n",
+		       cmd->input, cmd->parameter_amount, parameter_amount);
+		goto err;
+	}
+
+	return mfrc522_command_init(cmd->cmd, extra_data, extra_data_len);
```

Comment:
not a fan of the allocate, we could just copy the extra_data in the init function